File: Data Structures & Algorithms/construct-binary-tree-from-inorder-and-postorder-traversal/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int inIdx;
    int postIdx;

    TreeNode *dfs(vector<int>& postorder, vector<int>& inorder, int limit) {
        if(postIdx < 0) {
            return nullptr;
        }

        if(inorder[inIdx] == limit) {
            inIdx--;
            return nullptr;
        }

        TreeNode *root = new TreeNode(postorder[postIdx--]);
        root->right = dfs(postorder, inorder, root->val);
        root->left = dfs(postorder, inorder, limit);

        return root;
    }
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        postIdx = postorder.size() - 1;
        inIdx = inorder.size() - 1;

        return dfs(postorder, inorder, INT_MAX);
    }
};

File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recursiveInorder(self, root, ans):
        if root is None:
            return
        
        self.recursiveInorder(root.left, ans)
        ans.append(root.val)
        self.recursiveInorder(root.right, ans)

    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        ans = []

        self.recursiveInorder(root, ans)

        return ans
        